File: Algorithm-Swift/Algorithm-Swift/Inflearn/10_자릿수의합.swift
//
//  10_자릿수의합.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/07/28.
//

import Foundation

func digitSum(_ num: Int) -> Int {
    var num = num
    var sum = 0

    while num != 0 {
        sum += num % 10
        num /= 10
    }

    return sum
}

func solution10() {
    let _ = Int(readLine()!)!
    let nums = readLine()!.split(separator: " ").map { Int($0)! }
    var max = -99999999
    var result = 0

    nums.forEach { num in
        let sum = digitSum(num)
        if max < sum {
            max = sum
            result = num
        } else if max == sum && num > result {
            result = num
        }
    }

    print(result)
}

File: Algorithm-CPP/10_자릿수의합.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int digit_sum(int x) {
    int sum = 0;

    while(x != 0) {
        sum += x % 10;
        x /= 10;
    }

    return sum;
}

int main() {
    int n, result, maxValue = -99999999;
    int nums[100];
    
    cin >> n;

    for(int i=0; i<n; i++) {
        cin >> nums[i];
        int sum = digit_sum(nums[i]);

        if(sum > maxValue) {
            maxValue = digit_sum(nums[i]);
            result = nums[i];
        } else if(sum == maxValue && nums[i] > result) {
            result = nums[i];
        }
    }

    cout << result;

    return 0;
}